# Do traders actually make money being contrarian?

"Be greedy when others are fearful" is repeated constantly and tested rarely. It is a claim
about returns, and returns are measurable — if you line a sentiment reading up against the
realised profit of actual trades on the days that reading applied.

This notebook does that join and then asks three questions in order:

1. Does average profit differ between sentiment bands? **Yes, and significantly.**
2. Does that let you predict an individual trade? **No — ROC-AUC 0.53.**
3. How much of the effect is a handful of traders? **Most of it.**

Every stage is a call into `src/trader_sentiment/`, which is unit-tested and run in CI.

In [1]:
import pandas as pd

from trader_sentiment import analysis, clean, config, db, fear_greed, loaders, model, stats

pd.set_option("display.width", 120)

## 1. The data

Two sources. The trade export is a Hyperliquid account history; the sentiment index comes
straight from **alternative.me's public API**, so the project fetches its own inputs rather
than depending on a Google Drive link staying alive.

In [2]:
trades = loaders.load_trades(config.trades_csv())
sentiment = fear_greed.load_or_fetch(config.fear_greed_csv())

print(f"{len(trades):>10,} trades")
print(f"{len(sentiment):>10,} daily sentiment readings "
      f"({sentiment['date'].min()} to {sentiment['date'].max()})")

   211,224 trades
     3,135 daily sentiment readings (2018-02-01 to 2026-09-05)


### Join, then report coverage before dropping anything

A trade inherits the classification of the day it was placed. The join is a left join
deliberately, so unmatched trades stay visible long enough to be counted.

In [3]:
cleaned, coverage = clean.prepare(trades, sentiment)
print(f"unmatched: {coverage['missing']} of {coverage['total']:,} ({coverage['pct']}%)")
print(f"after the 3x IQR position filter: {len(cleaned):,} trades")

unmatched: 6 of 211,224 (0.0028%)
after the 3x IQR position filter: 167,331 trades


## 2. The headline: does the buyer or the seller earn more?

In [4]:
edge = analysis.contrarian_edge(cleaned)
edge.round(2)

,sentiment,buy_pnl,sell_pnl,buy_minus_sell,favours
0,Extreme Fear,55.67,24.20,31.47,BUY
1,Fear,64.81,37.73,27.08,BUY
2,Neutral,17.88,42.37,-24.49,SELL
3,Greed,35.01,39.32,-4.31,SELL
4,Extreme Greed,6.54,71.73,-65.18,SELL


**Positive means buyers earned more.** The pattern is monotonic: strongly positive under
fear, strongly negative under greed. That is the contrarian claim, and it is visible without
any modelling at all.

In [5]:
analysis.pnl_matrix(cleaned).round(2)

Side,BUY,SELL
sentiment,,
Extreme Fear,55.67,24.20
Fear,64.81,37.73
Neutral,17.88,42.37
Greed,35.01,39.32
Extreme Greed,6.54,71.73


## 3. Is that difference real, or is it noise?

The order of these tests is the argument:

- Profit is nowhere near normally distributed, which is **why** the comparison has to be
  rank-based rather than an ANOVA.
- Kruskal-Wallis says the bands differ *somewhere* — not where.
- Ten pairwise comparisons at 5% would throw a false positive about 40% of the time
  uncorrected, hence Bonferroni.

In [6]:
normality = stats.normality_by_group(cleaned)
print(f"bands that pass a normality test: {normality['normal'].sum()} of {len(normality)}")

result = stats.kruskal_across_groups(cleaned)
print(f"Kruskal-Wallis H = {result['statistic']:.2f}, p = {result['p_value']:.3g}")

pairs = stats.significant_pairs(stats.dunn_posthoc(cleaned))
print(f"pairs significant after Bonferroni: {len(pairs)} of 10")

bands that pass a normality test: 0 of 5
Kruskal-Wallis H = 685.66, p = 4.44e-147


pairs significant after Bonferroni: 10 of 10


## 4. The warehouse

The joined, cleaned trades go into SQLite once. Every question below is a `.sql` file that
CI executes, so the answers do not depend on this notebook being re-run correctly.

In [7]:
conn = db.connect(config.database_path())
print(f"{db.load_trades(conn, cleaned):,} rows written")
db.query(conn, "contrarian_edge.sql")

167,331 rows written


,sentiment,buy_pnl,sell_pnl,buy_minus_sell,favours
0,Extreme Fear,55.67,24.20,31.47,BUY
1,Fear,64.81,37.73,27.08,BUY
2,Neutral,17.88,42.37,-24.49,SELL
3,Greed,35.01,39.32,-4.31,SELL
4,Extreme Greed,6.54,71.73,-65.18,SELL


### Do fees eat the edge?

The original writeup listed "costs are excluded" as a limitation and left it there. It is
answerable.

In [8]:
db.query(conn, "fees_against_edge.sql")

,sentiment,trade_count,avg_pnl,avg_fee,avg_pnl_net_of_fee,fee_pct_of_pnl
0,Neutral,25684,29.45,1.1473,28.30,3.90
1,Fear,47161,50.83,1.8297,49.00,3.60
2,Greed,47002,37.27,1.3285,35.94,3.56
3,Extreme Fear,21526,39.43,1.2848,38.15,3.26
4,Extreme Greed,25958,41.13,0.7810,40.35,1.90


Fees consume roughly 2–4% of average profit, against an edge of +31 under Extreme Fear.
**The effect survives its own costs.**

### How much of this is a handful of traders?

This is the sharpest caveat in the project, and it was previously only stated in prose.

In [9]:
concentration = db.query(conn, "concentration_by_account.sql")
total_accounts = conn.execute("SELECT COUNT(DISTINCT account) FROM trades").fetchone()[0]

print(f"accounts in the dataset: {total_accounts}")
print(f"top 1  holds {concentration['pct_of_total_pnl'].iloc[0]:.1f}% of all profit")
print(f"top 5  hold {concentration['pct_of_total_pnl'].head(5).sum():.1f}%")
print(f"top 10 hold {concentration['pct_of_total_pnl'].head(10).sum():.1f}%")
concentration.head(5)

accounts in the dataset: 32
top 1  holds 23.5% of all profit
top 5  hold 62.6%
top 10 hold 83.9%


,account,trades,pnl,pct_of_total_pnl
0,0x083384f897ee0f19899168e3b1bec365f52a9012,3818,1600229.82,23.46
1,0x513b8629fe877bb581bf244e326a047b249c4ff1,12142,835309.06,12.24
2,0xbee1707d6b44d4d52bfe19e41f8a828645437aab,37817,722903.17,10.60
3,0xb1231a4a2dd02f2276fa3c5e2a2f3436e6bfed23,12891,560947.16,8.22
4,0x4acb90e786d897ecffb614dc822eb231b4ffb9f4,3361,553111.24,8.11


**167,331 trades, 32 accounts.** The trade count is a big number attached to a small
sample. "Traders earn more buying into fear" is much closer to "these few traders did".

## 5. Can it predict an individual trade?

A difference in *average* profit between bands is not the same claim as being able to call
a *single* trade, and conflating the two is how a backtest starts lying.

One filter matters enormously here. Roughly half the export's rows are opens and adds,
which carry `closed_pnl == 0` because no position was closed. Labelling those "not
profitable" turns the task into *"is this row a close?"* — a question about the file format,
which `Start Position` answers outright. Doing that scored ROC-AUC **0.94**, and none of it
was real.

In [10]:
closes = model.closing_trades(cleaned)
print(f"{len(cleaned):,} rows -> {len(closes):,} genuine closes")
print(f"base rate among closes: {(closes['Closed PnL'] > 0).mean():.4f}")

train_df, test_df = model.time_ordered_split(closes, train_frac=0.7)
x_train, y_train = model.build_features(train_df)
x_test, y_test = model.build_features(test_df)

scores = model.evaluate(model.train(x_train, y_train), x_test, y_test)
print(f"\nROC-AUC   {scores['roc_auc']:.4f}")
print(f"base rate {scores['base_rate']:.4f}")
print(f"accuracy  {scores['accuracy']:.4f}  (lift over base: "
      f"{scores['accuracy_lift_over_base']:+.4f})")

167,331 rows -> 80,522 genuine closes
base rate among closes: 0.8360



ROC-AUC   0.5310
base rate 0.8376
accuracy  0.5050  (lift over base: -0.3325)


**ROC-AUC 0.53.** Sentiment moves the average; it does not call the trade.

That is not a disappointing result, it is the precise one. The band-level effect is real
and significant, and it is still not a per-trade signal — which is exactly the distinction
that separates a finding from a trading strategy.

## What this shows

| question | answer |
|---|---|
| Do sentiment bands differ in average profit? | Yes — monotonic, and all 10 pairs survive Bonferroni |
| Does the edge survive fees? | Yes — costs are 2–4% of profit against a +31 edge |
| Can it predict an individual trade? | No — ROC-AUC 0.53 |
| How broad is the sample? | 32 accounts; the top one holds 23.5% of all profit |

## Limitations

- **Correlation, not causation.** Nothing here shows sentiment *caused* the difference.
- **Daily resolution.** A trade placed during a mid-day reversal carries the whole day's
  classification.
- **32 accounts.** The single largest caveat, quantified above.
- **One venue, one asset class.** Hyperliquid perpetuals.
- **No out-of-sample test of the strategy.** The contrarian rule is derived from the same
  data used to measure the effect, so it describes the past rather than predicting.
- **Slippage and funding are still excluded**, even though fees are now accounted for.